### PR TITLE
Add the pages.snippets.listSnippets event

### PR DESCRIPTION
**Pull request by tomaszstrojny**
Hello,

I wanted to remove a snippet from the list of the snippets and haven't found any way. I've decided to add the event because we could use it to add, remove, or edit the snippets. Makes much sense to me.

If the name or the docs are not sufficient, let me know.

Best regards,

Tomasz Strojny

### CODE REVIEW

**Comment by bennothommo** [comment]
@tomaszstrojny I feel that allowing access to the entire snippets property might be fraught with peril. It might be better to add some form of "removeSnippet" method and allow access to the snippet manager through the event instead, like how we do with the navigation events and the like.

What are your thoughts on that?

Also, any event that you want to add must have an event docblock like the following, which details the event and provides an example of its usage: https://github.com/rainlab/pages-plugin/blob/master/classes/Menu.php#L263-L295

**Comment by tomaszstrojny** [comment]
@bennothommo you're right. It may not be the best idea to allow access to the entire property. Although, this may be used to delete the entry as well as dynamically add or modify the snippet. Let's say I'm using a plugin that has hardcoded snippet name and description and I want to translate them without changing the third-party plugin?

In this case, if we add the `removeSnippet` we should also add something like `addSnippet` or give any other way to dynamically manage them.

I treat `Pages` plugin as a place for the client and as a developer I have to have a possibility to manage every piece of the functionality the client is getting. That's why, in my opinion, we don't have to touch the registered components, as they are accessible only for the developers (CMS section), but snippets are used in the "client space" which has to be fully customizable.

I haven't written the doc block as I was aware that this may not be the best idea. Once we come up with the solution I'll be happy to add it :)

What do you think about it?

**Comment by bennothommo** [comment]
@tomaszstrojny Yeah I think I'd prefer having specific `addSnippet` and `removeSnippet` methods to control the management of snippets through an event. Would you be willing to adjust your PR to do that for us? :)

**Comment by tomaszstrojny** [comment]
@bennothommo of course I will do that. Just let me know how would you prefer to have it done.

I would add the methods to the `SnippetManager` as public ones and make `$this` as a parameter of the event. But what event and when to fire them to make it work? I think that the place I've suggested is the most appropriate but let me know if you have any other ideas. The same applies to the name of the event. I'm not sure it will still make sense.

**Comment by bennothommo** [comment]
@tomaszstrojny here's my thoughts:

- Yep, I would simply make the event parameter the `SnippetManager` instance (ie. `$this`) as you've indicated.
- Add the two new methods to `SnippetManager` - `addSnippet` and `removeSnippet` - that will be available to the event *after* it has already loaded the theme and component snippets. (Pretty much where you have the event already).
- The line `$this->snippets = $snippets;` can be removed as the methods above will cover that already.
- I'd call the event `pages.snippets.listSnippets` to keep with the conventions of other events in the Pages plugin.

Feel free to let me know if you have any other questions.

**Comment by tomaszstrojny** [comment]
@bennothommo everything's clear. Take a look at the current version and send me your thoughts.

**Comment by mjauvin** [comment]
@tomaszstrojny some places in the code / comments still reference "snipet" singular...

**Comment by tomaszstrojny** [comment]
Whoops, all of them probably. Good spot. Give me a moment :)

File: classes/SnippetManager.php [hunk 1]
```diff
@@ -35,13 +35,58 @@ public function listSnippets($theme)
         $themeSnippets = $this->listThemeSnippets($theme);
         $componentSnippets = $this->listComponentSnippets();
 
-        $snippets = array_merge($themeSnippets, $componentSnippets);
+        $this->snippets = array_merge($themeSnippets, $componentSnippets);
+
+        /*
+         * @event pages.snippet.listSnippets
```
Comment (bennothommo) [inline classes/SnippetManager.php]: @tomaszstrojny snippets needs to be plural.
```suggestion
         * @event pages.snippets.listSnippets
```

File: classes/SnippetManager.php [hunk 2]
```diff
@@ -35,13 +35,58 @@ public function listSnippets($theme)
...
+         * Gives the ability to manage snippet list dynamically.
```
Comment (bennothommo) [inline classes/SnippetManager.php:42]:
```suggestion
         * Gives the ability to manage the snippet list dynamically.
```

File: classes/SnippetManager.php [hunk 3]
```diff
@@ -35,13 +35,58 @@ public function listSnippets($theme)
...
+         *
+         * Example usage to add a snippet to the list:
+         *
+         * Event::listen('pages.snippet.listSnippets', function(&$manager) {
```
Comment (bennothommo) [inline classes/SnippetManager.php]: @tomaszstrojny `$manager` does not need the pass-by-reference (`&`) character, as it is a class instance and is already passed by reference.
```suggestion
         * Event::listen('pages.snippets.listSnippets', function ($manager) {
```

File: classes/SnippetManager.php [hunk 4]
```diff
@@ -35,13 +35,58 @@ public function listSnippets($theme)
...
+         *     $snippet = new \RainLab\Pages\Classes\Snippet();
+         *     $snippet->initFromComponentInfo('\Example\Plugin\Components\ComponentClass', 'snippetCode');
+         *     $manager->addSnippet($snippet);
+         * });
+         *
+         * Example usage to remove a snippet from the list:
+         *
+         * Event::listen('pages.snippet.listSnippets', function(&$manager) {
```
Comment (bennothommo) [inline classes/SnippetManager.php]:
```suggestion
         * Event::listen('pages.snippets.listSnippets', function ($manager) {
```

File: classes/SnippetManager.php [hunk 5]
```diff
@@ -35,13 +35,58 @@ public function listSnippets($theme)
...
+         *     $manager->removeSnippet('snippetCode');
+         * });
+         */
+        Event::fire('pages.snippet.listSnippets', [&$this]);
```
Comment (bennothommo) [inline classes/SnippetManager.php]:
```suggestion
        Event::fire('pages.snippets.listSnippets', [$this]);
```

File: classes/SnippetManager.php [hunk 6]
```diff
@@ -35,13 +35,58 @@ public function listSnippets($theme)
...
+    {
+        $newSnippets = [];
+        foreach ($this->snippets as $snippet) {
+            if ($snippet->code !== $snippetCode) {
+                $newSnippets[] = $snippet;
+            }
+        }
 
-        return $snippets;
+        $this->snippets = $newSnippets;
```
Comment (bennothommo) [inline classes/SnippetManager.php]: @tomaszstrojny This method can be simplified to the following:

```php
$this->snippets = array_filter($this->snippets, function ($snippet) use ($snippetCode) {
    return $snippet->code !== $snippetCode;
});
```

File: docs/documentation.md [hunk 7]
```diff
@@ -347,20 +347,21 @@ A same component can be registered with registerPageSnippets() and registerCompo
...
-            if ($snippet->code !== 'unneededSnippet') {
-                $newSnippets[] = $snippet;
-            }
-        }
-        $snippets = $newSnippets;
+If you want to dynamically extend the list of the snippets you can bind to the `pages.snippets.listSnippets` event.
+
+An example usage to add a snippet to the list:
+
+    Event::listen('pages.snippet.listSnippets', function(&$manager) {
```
Comment (bennothommo) [inline docs/documentation.md]:
```suggestion
    Event::listen('pages.snippets.listSnippets', function ($manager) {
```

File: docs/documentation.md [hunk 8]
```diff
@@ -347,20 +347,21 @@ A same component can be registered with registerPageSnippets() and registerCompo
...
+        $snippet = new \RainLab\Pages\Classes\Snippet();
+        $snippet->initFromComponentInfo('\Example\Plugin\Components\ComponentClass', 'snippetCode');
+        $manager->addSnippet($snippet);
+    });
+
+An example usage to remove a snippet from the list:
+
+    Event::listen('pages.snippet.listSnippets', function(&$manager) {
```
Comment (bennothommo) [inline docs/documentation.md]:
```suggestion
    Event::listen('pages.snippets.listSnippets', function ($manager) {
```

**Review by bennothommo — CHANGES_REQUESTED** [review]

**Comment by bennothommo** [comment]
@tomaszstrojny I realised that you fixed up the plurals, but you'll need to remove the pass-by-reference calls as well.

**Comment by daftspunk** [comment]
Thanks @tomaszstrojny !

**Pull request merged**